src/sentry/models/release.py

```diff
@@ -381,61 +381,67 @@ def set_commits(self, commit_list):
-                            author.update(name=data['author_name'])
+                            values=author_data)
+                        if not created:
+                            author = CommitAuthor.objects.get(
+                                organization_id=self.organization_id,
+                                email=author_email)
+                        authors[author_email] = author
                     else:
                         author = authors[author_email]
 
-                    defaults = {
+                    commit_data = {
                         'message': data.get('message'),
-                        'author': author,
                         'date_added': data.get('timestamp') or timezone.now(),
                     }
-                    commit, created = Commit.objects.get_or_create(
+                    # If we didn't get an author don't overwrite an existing one.
+                    if author is not None:
+                        commit_data['author'] = author
+
+                    commit, created = Commit.objects.create_or_update(
                         organization_id=self.organization_id,
                         repository_id=repo.id,
                         key=data['id'],
-                        defaults=defaults,
-                    )
+                        values=commit_data)
+                    if not created:
+                        commit = Commit.objects.get(
+                            organization_id=self.organization_id,
+                            repository_id=repo.id,
+                            key=data['id'])
+
                     if author is None:
                         author = commit.author
 
                     commit_author_by_commit[commit.id] = author
 
                     patch_set = data.get('patch_set', [])
-
                     for patched_file in patch_set:
-                        CommitFileChange.objects.get_or_create(
-                            organization_id=self.organization.id,
+                        try:
+                            CommitFileChange.objects.create(
+                                organization_id=self.organization.id,
+                                commit=commit,
+                                filename=patched_file['path'],
+                                type=patched_file['type'],
+                            )
+                        except IntegrityError:
+                            pass
+
+                    try:
+                        ReleaseCommit.objects.create(
+                            organization_id=self.organization_id,
+                            release=self,
                             commit=commit,
-                            filename=patched_file['path'],
-                            type=patched_file['type'],
+                            order=idx,
                         )
+                    except IntegrityError:
```

Comment:
shouldn't this not be happening given the lock? if it is happening, it's likely due to a bug and we maybe shouldn't just silence the integrity errors

Reply:
Should we be increasing the amount of time we acquire the lock for? (Maybe we could capture some stats on how long this segment of code is taking and validate that this is actually the cause, and not some other bug with the lock)

Reply:
I can add metrics to track duration of the locked block and see if it could be lock expiration.

Reply:
I took a look at the metrics in datadog and either the metrics for the failing situations are getting elided by sampling or I'm wrong with how this problem comes up.

Reply:
Weird, this seemed like a good candidate for how this is happening. I went through a couple days of data in Datadog by hour and saw a couple stats in the 5-6 range, but nothing close to the 10 second mark. It's possible that it's getting dropped off though...